lang/cs/Org.Apache.REEF.IMRU/OnREEF/Driver/IMRUDriver.cs

```diff
@@ -98,181 +128,462 @@ internal sealed class IMRUDriver<TMapInput, TMapOutput, TResult, TPartitionType>
+        /// <summary>
+        /// IFailedEvaluator handler. It specifies what to do when an evaluator fails.
+        /// If we get all completed tasks then ignore the failure. Otherwise, take the following actions based on the system state: 
+        /// Case WaitingForEvaluator
+        ///     This happens in the middle of submitting contexts. We just need to remove the failed evaluator 
+        ///     from EvaluatorManager and remove associated active context, if any, from ActiveContextManager
+        ///     then checks if the system is recoverable. If yes, request another Evaluator 
+        ///     If not recoverable, set system state to Fail then execute Fail action
+        /// Case SubmittingTasks/TasksRunning
+        ///     This happens either in the middle of Task submitting or all the tasks are running
+        ///     Changes the system state to ShuttingDown
+        ///     Removes Evaluator and associated context from EvaluatorManager and ActiveContextManager
+        ///     Removes associated task from running task if it was running and change the task state to TaskFailedByEvaluatorFailure
+        ///     Closes all the other running tasks
+        ///     Try to recover in case it is the last failure received
+        /// Case ShuttingDown
+        ///     This happens when we have received either FailedEvaluator or FailedTask, some tasks are running some are in closing.
+        ///     Removes Evaluator and associated context from EvaluatorManager and ActiveContextManager
+        ///     Removes associated task from running task if it was running, changes the task state to ClosedTask if it was waiting for close
+        ///     otherwise changes the task state to FailedTaskEvaluatorError
+        ///     Try to recover in case it is the last failure received
+        /// Other cases - not expected 
+        /// </summary>
+        /// <param name="failedEvaluator"></param>
+        public void OnNext(IFailedEvaluator failedEvaluator)
+        {
+            lock (_lock)
+            {
+                if (_taskManager.AreAllTasksCompleted())
+                {
+                    Logger.Log(Level.Verbose, "Evaluator with Id: {0} failed but IMRU task is completed. So ignoring.", failedEvaluator.Id);
+                    return;
+                }
+                Logger.Log(Level.Info, "Evaluator with Id: {0} failed with Exception: {1} in retry # {2}.", failedEvaluator.Id, failedEvaluator.EvaluatorException, _numberOfRetryForFaultTolerant);
+
+                var isMaster = _evaluatorManager.IsMasterEvaluatorId(failedEvaluator.Id);
+                _evaluatorManager.RecordFailedEvaluator(failedEvaluator.Id);
+                _contextManager.RemoveFailedContextInFailedEvaluator(failedEvaluator);
+
+                switch (_systemState.CurrentState)
+                {
+                    case SystemState.WaitingForEvaluator:
+                        if (!_evaluatorManager.ReachedMaximumNumberOfEvaluatorFailures())
```

Comment:
yes, if the manager "knows" maximum number of retries it can decide when to trigger nonRecoverable error.  

---
In reply to: [67057535](https://github.com/apache/reef/pull/1033#discussion_r67057535) [](ancestors = 67057535)